src/dot-protocol/merlin_dot_protocol.mli

```diff
@@ -74,19 +74,42 @@ end
 
 type directive = Directive.Processed.t
 
-module Commands : sig
-  type t = File of string | Halt | Unknown
-
-  val read_input : in_channel -> t
-  val send_file : out_channel:out_channel -> string -> unit
-end
-
 type read_error =
   | Unexpected_output of string
   | Csexp_parse_error of string
 
-(** [read inc] reads one csexp from the channel [inc] and returns the list of
-  directives it represents *)
-val read : in_channel:in_channel -> (directive list, read_error) Merlin_utils.Std.Result.t
+module type Chan = sig
+  module IO : sig
+    type 'a t
+
+    module O : sig
+      val ( let+ ) : 'a t -> ('a -> 'b) -> 'b t
+    end
+  end
+
+  type in_chan
+  type out_chan
+
+  val read : in_chan -> (Csexp.t, string) result IO.t
+
+  val write : out_chan -> Csexp.t -> unit IO.t
+end
+
+module Make (Chan : Chan) : sig
+  (** [read inc] reads one csexp from the channel [inc] and returns the list of
+      directives it represents *)
+  val read : Chan.in_chan -> (directive list, read_error) Merlin_utils.Std.Result.t Chan.IO.t
+
+  val write : Chan.out_chan -> directive list -> unit Chan.IO.t
+
+  module Commands : sig
+    type t = File of string | Halt | Unknown
+    val read_input : Chan.in_chan -> t Chan.IO.t
+
+    val send_file : Chan.out_chan -> string -> unit Chan.IO.t
+
+    val halt : Chan.out_chan -> unit Chan.IO.t
+  end
+end
 
-val write : out_channel:out_channel -> directive list -> unit
+module StdChan : Chan with type in_chan = in_channel and type out_chan = out_channel and type 'a IO.t = 'a
```

Comment:
I think we should instantiate the functor here so that we can use it directly in Merlin.

Anyway, configuration loading is tightly tied to `merlin-lib` right now, we should make it more modular at some point. For exemple, `ocaml-lsp` will still need to re-implement `mconfig_dot.ml` after this patch: https://github.com/ocaml/ocaml-lsp/blob/master/ocaml-lsp-server/src/merlin_config.ml 

(but at least with this patch they can stop vendoring)

Reply:
Sure, I have created `Merlin_dot_protocol.Std` module.